src/group_theory/perm/cycle/basic.lean

```diff
@@ -605,6 +619,166 @@ end
 
 end is_cycle
 
+/-! ### `is_cycle_on` -/
+
+section is_cycle_on
+variables {f g : perm α} {s t : set α} {a b x y : α}
+
+/-- A permutation is a cycle on `s` when any two points of `s` are related by repeated application
+of the permutation. -/
+def is_cycle_on (f : perm α) (s : set α) : Prop :=
+set.bij_on f s s ∧ ∀ ⦃x⦄, x ∈ s → ∀ ⦃y⦄, y ∈ s → f.same_cycle x y
+
+lemma is_cycle_on_empty :  f.is_cycle_on ∅ := by simp [is_cycle_on, set.bij_on_empty]
+
+lemma is_cycle_on.subsingleton (h : (1 : perm α).is_cycle_on s) : s.subsingleton :=
+λ x hx y hy, same_cycle_one.1 $ h.2 hx hy
+
+@[simp] lemma is_cycle_on_singleton : f.is_cycle_on {a} ↔ f a = a :=
+by simp [is_cycle_on, same_cycle.rfl]
+
+lemma is_cycle_on_of_subsingleton [subsingleton α] (f : perm α) (s : set α) : f.is_cycle_on s :=
+⟨s.bij_on_of_subsingleton _, λ x _ y _, (subsingleton.elim x y).same_cycle _⟩
+
+lemma is_cycle_on.inv (h : f.is_cycle_on s) : f⁻¹.is_cycle_on s :=
+⟨h.1.perm_inv, λ x hx y hy, (h.2 hx hy).inv⟩
+
+lemma is_cycle_on.of_inv (h : f⁻¹.is_cycle_on s) : f.is_cycle_on s :=
+by { convert h.inv, rw inv_inv }
+
+@[simp] lemma is_cycle_on_inv : f⁻¹.is_cycle_on s ↔ f.is_cycle_on s :=
+⟨is_cycle_on.of_inv, is_cycle_on.inv⟩
+
+lemma is_cycle_on.conj (h : f.is_cycle_on s) : (g * f * g⁻¹).is_cycle_on ((g⁻¹ : perm α) ⁻¹' s) :=
```

Comment:
Hmm, not sure. We tend to prefer preimages where necessary.

Maybe `image_inv` and `preimage_inv` shouldn't be simp lemmas, but I quite like them like that.

Reply:
I agree that preimages are nicer in general, but getting rid of inverses seems mathematically like the more obvious way to simplify. And since `g` is a permutation, there still is a nice formulation of the image that doesn't involve existentials. So I think my suggestion really is the right way to go.